### PR TITLE
Add support for file write/upload operations with `HfHubRepository`

### PR DESCRIPTION
**Pull request by shadeMe**
## Description

This PR introduces support for uploading files to Hugging Face Hub repositories through the current `RepositoryFile` API. In the case of `fsspec` repositories, we support implicit streaming of data using buffered IO. However, this is not possible with HF repositories due its usage of Git as a backing store. So, we need to allocate and write to files on the host's local storage device before uploading them as a Git commit operation to the remote repo.

The current implementation hides this detail by using a proxy to represent the remote file that we want to write to. The `HfHubFile` class wraps either a locally cached file from a HF repo or a temporary file on the local storage. The user can open and write to the latter like they would with any (`fsspec`) file - its contents will be uploaded to the repo as soon as the file handle is closed.

Furthermore, we also introduce the concept of transactional file writes using a context manager. This lets the user batch multiple file operations that get uploaded to the repo as a single commit. The `fsspec` implementation of transactions will be implemented in a follow-up PR.

Other changes:
- `Repository.file` returns a lazily-loaded file, i.e., its existence is only checked when the file is opened using the `open` method (or when the `exists` method is called).
- Fixed a bug in `FsspecRepository.open` that called `unstrip_protocol` on a `None` object.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.

### CODE REVIEW

File: curated_transformers/repository/hf_hub.py [hunk 1]
```diff
@@ -31,26 +112,201 @@ def __init__(self, name: str, *, revision: str = "main"):
...
+            operations=commit_ops,
+            commit_message=commit_message,
+            run_as_future=False,
+        )
+
+    def _release_temp_files(self):
+        for temp_file in self._file_mappings.values():
+            if not temp_file.closed:
+                temp_file.close()
+            os.remove(temp_file.name)
```
Comment (danieldk) [inline curated_transformers/repository/hf_hub.py:199]: Same here, these can raise, if that happens, we should continue removing temporary files.

File: curated_transformers/tests/repository/test_hf_hub.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import pytest
+
+from curated_transformers.repository import HfHubRepository
+
+TEST_LINES = ["Line 1", "Line 2 🐍", "🤖" "\n\n", ""]
+
+
+@pytest.mark.slow
```
Comment (danieldk) [inline curated_transformers/tests/repository/test_hf_hub.py:8]: I think we need another mark. We still want people to be able to run slow tests, even if they don't have access to our repo.

File: curated_transformers/tokenizers/tokenizer.py [hunk 3]
```diff
@@ -337,8 +338,16 @@ def from_hf_hub_to_cache(
...
     def from_repo(cls: Type[Self], repo: Repository) -> Self:
+        def validate_tokenizer_file(tokenizer_file: RepositoryFile):
+            # Since repository files are loaded lazily, we need to
+            # open them to check their validity. Will throw if invalid.
+            with tokenizer_file.open(mode="rb") as f:
+                pass
+
         repo = TokenizerRepository(repo)
         tokenizer_file = repo.tokenizer_json()
+        validate_tokenizer_file(tokenizer_file)
```
Comment (danieldk) [inline curated_transformers/tokenizers/tokenizer.py:349]: Hmmm, this does make an additional request/has some overhead.

**Review by danieldk — COMMENTED** [review]
🎉 

Added some comments

File: curated_transformers/repository/hf_hub.py [hunk 4]
```diff
@@ -31,26 +112,201 @@ def __init__(self, name: str, *, revision: str = "main"):
...
+    def repo(self) -> Repository:
+        return self._repo
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        if exc_type is None:
+            self._upload_temp_files()
+        self._release_temp_files()
```
[the remark this answers is not in the document]
Comment (shadeMe) [reply curated_transformers/repository/hf_hub.py]: Done.

File: curated_transformers/repository/hf_hub.py (hunk 1 above)
Comment (shadeMe) [reply curated_transformers/repository/hf_hub.py]: Done.

File: curated_transformers/tests/repository/test_hf_hub.py (hunk 2 above)
Comment (shadeMe) [reply curated_transformers/tests/repository/test_hf_hub.py]: Added a new marker for upload tests.

File: curated_transformers/tokenizers/tokenizer.py (hunk 3 above)
Comment (shadeMe) [reply curated_transformers/tokenizers/tokenizer.py]: Rewrote the initializer to use a single request.

File: curated_transformers/repository/fsspec.py [hunk 5]
```diff
@@ -97,15 +107,47 @@ def __init__(
...
+
     @property
     def _protocol(self) -> str:
         return self.fs.unstrip_protocol(self.repo_path)
+
+
+class FsspecTransactionContext(TransactionContext):
+    """
+    `fsspec`_ transaction context manager.
+    TODO: Implement/currently a noop
```
Comment (rmitsch) [inline curated_transformers/repository/fsspec.py:128]: Should the TODO be in here? How is this a noop?

File: curated_transformers/repository/hf_hub.py [hunk 6]
```diff
@@ -11,6 +14,84 @@
...
+        super().__init__()
+
+        self._repo = repo
+        self._remote_path = path
+        self._cached_path = lookup_file_in_cache(repo.name, repo.revision, path)
+
+    def _validate_model_repo(self):
+        api = HfApi()
+        # This will raise RepositoryNotFoundError/RevisionNotFoundError upon failure.
+        _ = api.model_info(repo_id=self._repo.name, revision=self._repo.revision)
```
Comment (rmitsch) [inline curated_transformers/repository/hf_hub.py:44]: Are you setting `_` on purpose instead of
```suggestion
        HfApi().model_info(repo_id=self._repo.name, revision=self._repo.revision)
```
?

File: curated_transformers/repository/hf_hub.py [hunk 7]
```diff
@@ -31,26 +112,207 @@ def __init__(self, name: str, *, revision: str = "main"):
...
+            self._release_temp_files()
+
+    def _upload_temp_files(self):
+        if len(self._file_mappings) == 0:
+            return
+
+        # Flush buffers.
+        for temp_file in self._file_mappings.values():
+            if not temp_file.closed:
+                temp_file.flush()
```
Comment (rmitsch) [inline curated_transformers/repository/hf_hub.py:178]: Why is this necessary?

File: curated_transformers/repository/hf_hub.py [hunk 8]
```diff
@@ -31,26 +112,207 @@ def __init__(self, name: str, *, revision: str = "main"):
...
+        # Collate all files as a single commit.
+        commit_ops = [
+            CommitOperationAdd(path_in_repo=remote_path, path_or_fileobj=temp_file.name)
+            for remote_path, temp_file in self._file_mappings.items()
+        ]
+        filenames = "\n".join(self._file_mappings.keys())
+        commit_message = f"Uploading files:\n{filenames}"
+
+        api = HfApi()
+        commit_info = api.create_commit(
```
Comment (rmitsch) [inline curated_transformers/repository/hf_hub.py:189]:
```suggestion
        HfApi().api.create_commit(
```
`commit_info` isn't used, right?

**Review by rmitsch — COMMENTED** [review]
I'm not familar with the codebase yet, so my comments are somewhat superficial. Overall this looks reasonable to me. I'll do a second pass after the first batch of comments has been addressed.

File: curated_transformers/repository/file.py [hunk 9]
```diff
@@ -40,6 +43,14 @@ def path(self) -> Optional[str]:
         """
         ...
 
+    @abstractmethod
+    def exists(self) -> bool:
+        """
+        Returns if the file exists. This can cause the file
+        to be cached locally.
+        """
```
[the remark this answers is not in the document]
Comment (shadeMe) [reply curated_transformers/repository/file.py:51]: We use [sphinx](https://www.sphinx-doc.org/en/master/) for our docs in this codebase, so the return type is automatically generated. We also don't explicitly mention the return value for property-like methods.

File: curated_transformers/repository/fsspec.py (hunk 5 above)
Comment (shadeMe) [reply curated_transformers/repository/fsspec.py:128]: Yeah, it's intended. The transaction semantics of `fsspec` are different from how we handle it for HF repos, so we need to think about it a bit more. The stub class is just there to aid the test suite/CI.

File: curated_transformers/repository/hf_hub.py (hunk 6 above)
Comment (shadeMe) [reply curated_transformers/repository/hf_hub.py:44]: Yeah, it's to indicate that we are intentionally discarding the return value.

File: curated_transformers/repository/hf_hub.py (hunk 7 above)
Comment (shadeMe) [reply curated_transformers/repository/hf_hub.py:178]: More of a sanity check as the user can consume the `IO`-like object outside of a `with` context.

File: curated_transformers/repository/hf_hub.py (hunk 8 above)
Comment (shadeMe) [reply curated_transformers/repository/hf_hub.py:189]: True, but I'll `_` it to indicate that we're discarding the result (we might need it later).
